src/stream.ts

```diff
@@ -107,203 +99,123 @@ export class YamuxStream implements Stream {
     this.epochStart = Date.now()
     this.getRTT = init.getRTT
 
-    this.abortController = new AbortController()
-
     this.sendFrame = init.sendFrame
-    this.onStreamEnd = init.onStreamEnd
-
-    this.sourceInput = pushable({
-      onEnd: (err?: Error) => {
-        if (err != null) {
-          this.log?.error('stream source ended id=%s', this._id, err)
-        } else {
-          this.log?.trace('stream source ended id=%s', this._id)
-        }
 
-        this.closeRead()
-      }
+    this.source = each(this.source, () => {
+      this.sendWindowUpdate()
     })
-
-    this.source = this.createSource()
-
-    this.sink = async (source: Source<Uint8Array | Uint8ArrayList>): Promise<void> => {
-      if (this.writeState !== HalfStreamState.Open) {
-        throw new Error('stream closed for writing')
-      }
-
-      source = abortableSource(source, this.abortController.signal, { returnOnAbort: true })
-
-      try {
-        for await (let data of source) {
-          // send in chunks, waiting for window updates
-          while (data.length !== 0) {
-            // wait for the send window to refill
-            if (this.sendWindowCapacity === 0) await this.waitForSendWindowCapacity()
-
-            // send as much as we can
-            const toSend = Math.min(this.sendWindowCapacity, this.config.maxMessageSize - HEADER_LENGTH, data.length)
-            this.sendData(data.subarray(0, toSend))
-            this.sendWindowCapacity -= toSend
-            data = data.subarray(toSend)
-          }
-        }
-      } catch (e) {
-        this.log?.error('stream sink error id=%s', this._id, e)
-      } finally {
-        this.log?.trace('stream sink ended id=%s', this._id)
-        this.closeWrite()
-      }
-    }
   }
 
-  private async * createSource (): AsyncGenerator<Uint8ArrayList> {
-    try {
-      for await (const val of this.sourceInput) {
-        this.sendWindowUpdate()
-        yield val
-      }
-    } catch (err) {
-      const errCode = (err as { code: string }).code
-      if (errCode !== ERR_STREAM_ABORT) {
-        this.log?.error('stream source error id=%s', this._id, err)
-        throw err
-      }
-    }
-  }
-
-  close (): void {
-    this.log?.trace('stream close id=%s', this._id)
-    this.closeRead()
-    this.closeWrite()
-  }
-
-  closeRead (): void {
-    if (this.state === StreamState.Finished) {
-      return
-    }
-
-    if (this.readState !== HalfStreamState.Open) {
-      return
-    }
-
-    this.log?.trace('stream close read id=%s', this._id)
-
-    this.readState = HalfStreamState.Closed
-
-    // close the source
-    this.sourceInput.end()
+  /**
+   * Send a message to the remote muxer informing them a new stream is being
+   * opened
+   */
+  async sendNewStream (): Promise<void> {
```

Comment:
I guess we don't need this because we sent a SYN flag along with the first data to be sent? maybe worth commenting here

Reply:
Sort of, yeah - the stream is [opened by the `.newStream` method in the muxer](https://github.com/ChainSafe/js-libp2p-yamux/blob/master/src/muxer.ts#L191) before any data is sent.

Was this intentional?  If the `sendWindowUpdate` call is removed from the muxer it can switch to the behaviour of the other muxers which is to lazily signal opening the stream before the first data message is sent.

Reply:
I've updated the method comment to this effect.